src/libraries/Microsoft.Extensions.Hosting/src/HostApplicationBuilder.cs

```diff
@@ -98,6 +100,47 @@ public HostApplicationBuilder(HostApplicationBuilderSettings? settings)
                 HostingHostBuilderExtensions.AddCommandLineConfig(Configuration, settings.Args);
             }
 
+            Initialize(settings, out _hostBuilderContext, out _environment, out _logging);
+
+            ServiceProviderOptions? serviceProviderOptions = null;
+            if (!settings.DisableDefaults)
+            {
+                HostingHostBuilderExtensions.ApplyDefaultAppConfiguration(_hostBuilderContext, Configuration, settings.Args);
+                HostingHostBuilderExtensions.AddDefaultServices(_hostBuilderContext, Services);
+                serviceProviderOptions = HostingHostBuilderExtensions.CreateDefaultServiceProviderOptions(_hostBuilderContext);
+            }
+
+            _createServiceProvider = () =>
+            {
+                // Call _configureContainer in case anyone adds callbacks via HostBuilderAdapter.ConfigureContainer<IServiceCollection>() during build.
+                // Otherwise, this no-ops.
+                _configureContainer(Services);
+                return serviceProviderOptions is null ? Services.BuildServiceProvider() : Services.BuildServiceProvider(serviceProviderOptions);
+            };
+        }
+
+        internal HostApplicationBuilder(HostApplicationBuilderSettings? settings, bool empty)
+        {
+            Debug.Assert(empty, "should only be called with empty: true");
+
+            settings ??= new HostApplicationBuilderSettings();
+            Configuration = settings.Configuration ?? new ConfigurationManager();
+
+            HostingHostBuilderExtensions.AddCommandLineConfig(Configuration, settings.Args);
```

Comment:
Nit: Should we just move this into `Initialize` and call `hostConfigBuilder.AddEnvironmentVariables(prefix: "DOTNET_");` and get rid of `AddDefaultHostConfigurationSources`?

Reply:
Done.